### PR TITLE
User info (backend and frontend part)

### PR DESCRIPTION
**Pull request by RazSKg**

### CODE REVIEW

File: easyrest/models/user.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    The data model of "users" table
+    Defines data structure of "users" table
+    """
+    __tablename__ = 'users'
+    id = Column(Integer, primary_key=True)
+    name = Column(Text)
+    email = Column(Text)
+    phone_number = Column(Text)
+    birth_date = Column(Date)
+    role_id = Column(Integer)  # now it is without foreign key relations
```
Comment (serge30) [inline easyrest/models/user.py]: Common practice to use special todo comments for such things:
```
# TODO: Add relations
```

File: easyrest/scripts/json_helpers.py [hunk 2]
```diff
@@ -1,7 +1,17 @@
 """Script contains helper functions to work with json such as:
     wrap(): wraper to make response acording to style
+    date_time_normalize(): function, which normaloze date_time object for proper json serialization
 """
 
+import datetime
+
+
+def date_time_normalize(date_time_object):
```
Comment (serge30) [inline easyrest/scripts/json_helpers.py]: Docstring.

File: easyrest/views/user_info_controller.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                {
+                }
+            ]
+        If user is unauthorized - throw 403:
+    """
+    # in this place must be user authorization status check
+    # but now is simple User with Id = 1
+    user_id = 1
+    user = request.dbsession.query(User).get(user_id)
+    # debug (will replace when authorization will be done
```
Comment (serge30) [inline easyrest/views/user_info_controller.py]: Please use TODO comments.

File: easyrest/views/user_info_controller.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    # user = None
+    if user is None:
+        raise HTTPForbidden("403")
+    else:
+        user_dict = user.as_dict()
+        user_dict = {key: date_time_normalize(value) for key, value in user_dict.items()}
```
Comment (serge30) [inline easyrest/views/user_info_controller.py]: In pull #8  Max fixes `as_dict` method to handle datetimes, so probably you will not need this.

**Review by serge30 — CHANGES_REQUESTED** [review]

File: easyrest/views/user_info_controller.py (hunk 4 above)
Comment (RazSKg) [reply easyrest/views/user_info_controller.py]: Yes, i saw it. if you don't mind - I will remove this after Max will merge his branch with "develop". Then I will use his mechanism. Ok?

**Comment by RazSKg** [comment]
also added function, which redirect fo SignUp page when Forbidden.

File: easyrest/views/user_info_controller.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        body = wrap([user_dict])
```
Comment (viktormovchaniuk) [inline easyrest/views/user_info_controller.py]: look like you don't need wrap your userdict into brackets

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import { Redirect } from 'react-router';
+import {withStyles} from '@material-ui/core/styles';
+import UserInfo from "./UserInfo";
+import UserAvatar from "./UserAvatar";
+
+const styles = {
+    forDiv: {
+        flexGrow: 1,
```
Comment (viktormovchaniuk) [inline frontend/src/components/UserInfo/UserAvatarWithInfo.js]: Don't need `flex-grow` property here. By default flex container stretch to fill the space

File: frontend/src/components/UserInfo/UserAvatar.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import Typography from '@material-ui/core/Typography';
+
+const styles = {
+    card: {
+        flexGrow: 1,
+        maxWidth: 345,
+        width: "50%",
+        float: "left",
+        margin: 16,
+
```
Comment (viktormovchaniuk) [inline frontend/src/components/UserInfo/UserAvatar.js]: You don't really need `float` property almost in all cases, however your `float` isn't work here because your element inside `flex` container
Please for positioning  your elements use `<Grid/>` component or at least `flex-box` model

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        fetch('http://localhost:6543/profile')
+            .then(response => response.json())
+            .then(data => this.setState({userInfo: data.data[0], isForbidden: data.error.toLowerCase().search("forbidden") === -1 ? false: true}))
+            .catch(err=>console.log(err))
+    }
+
+    render() {
+        const {classes} = this.props;
+        const {userInfo, isForbidden} = this.state;
+        console.log(isForbidden);
```
Comment (viktormovchaniuk) [inline frontend/src/components/UserInfo/UserAvatarWithInfo.js]: don't commit debugging code

File: frontend/src/components/UserInfo/UserInfo.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    root: {
+        flexGrow: 1,
+        maxWidth: "50%",
+    },
+    table: {
+
+    },
+    cardStyle: {
+        width: "100%",
+        float: "left",
```
Comment (viktormovchaniuk) [inline frontend/src/components/UserInfo/UserInfo.js]: floats

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        if (isForbidden){
```
Comment (viktormovchaniuk) [inline frontend/src/components/UserInfo/UserAvatarWithInfo.js]: as i understood it is temporary, right?

Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: all logs I removed (but didn't commit yet) - I forgot remove them before commit.

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js (hunk 8 above)
Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: all logs I removed (but didn't commit yet) - I forgot remove them before commit.

File: easyrest/views/user_info_controller.py (hunk 5 above)
Comment (RazSKg) [reply easyrest/views/user_info_controller.py]: but wrap function data paramether is "data (list): list of dictionaries representing models output". I must do that.

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js (hunk 6 above)
Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: Ok. Thanks!

File: frontend/src/components/UserInfo/UserAvatar.js (hunk 7 above)
Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatar.js]: Thanks for info. I will use Grid.

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js (hunk 10 above)
Comment (viktormovchaniuk) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: i mean isForbidden cheeking, not console.log()

Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: If isForbidden is "true" - I return Redirection, else I return false. It influence in "UserAvatarWithInfo.js" in render() method

File: frontend/src/components/UserInfo/UserInfo.js (hunk 9 above)
Comment (RazSKg) [reply frontend/src/components/UserInfo/UserInfo.js]: fixed.

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js (hunk 10 above)
Comment (RazSKg) [reply frontend/src/components/UserInfo/UserAvatarWithInfo.js]: Sorry, this is outdated code. Those changes, which you describe - wasn`t presented in yesterday`s commit.

File: easyrest/views/user_info_controller.py (hunk 5 above)
Comment (serge30) [reply easyrest/views/user_info_controller.py]: I believe, description to this param is misleading. It doesn't require list here. Please send a record instead of list, and fix doctsring for `wrap` function.

File: easyrest/views/user_info_controller.py (hunk 4 above)
Comment (serge30) [reply easyrest/views/user_info_controller.py]: Ok

File: easyrest/scripts/insert_example_data.py [hunk 11]
```diff
@@ -81,5 +83,19 @@ def fill_db(session):
 
         Rest_models.append(rest_model)
 
+    # add users
+    for myIter in range(menu_item_number):
```
Comment (serge30) [inline easyrest/scripts/insert_example_data.py]: PEP8 for variable name. Anyway `iter` is bad suffix here as iterator is object we may iterate, instead of particular record. I'd use just `i` or `index` here.

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        const headers = new Headers({
+            'Content-Type': 'application/json',
+            'Authorization': this.state.token
+        });
+
+
+        fetch('http://localhost:6543/profile', {method: "GET", headers})
+            .then(response => response.json())
+            .then(data => this.setState({userInfo: data.data[0], success: data.success, error: data.error}))
+            .catch(err=>console.log(err))
```
Comment (serge30) [inline frontend/src/components/UserInfo/UserAvatarWithInfo.js]: Most of users don't know nothing about browser console. You need to let them now, that something went wrong.

File: frontend/src/Service/NeedAuthorization.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import { Redirect } from 'react-router';
+
+
+export function redirectToSignUp(error) {
+
+    /*this function checks error message for Forbidden cause.
+    If it present - return Redirect for LogIn page, else - return false*/
+
+    const isForbidden = ("" + error).toLowerCase().search("forbidden") !== -1;
```
Comment (serge30) [inline frontend/src/Service/NeedAuthorization.js]: I'm not sure about this approach, it's unsafe as you cannot be sure that other errors will not contain "forbidden" word. You'd better add some field to response dict along with `success` and `error` which indicate unauthorized access, or use proper HTTP status code 403.

**Review by serge30 — CHANGES_REQUESTED** [review]

**Comment by RazSKg** [comment]
Made corrections according to Your remarks. Add http status code for NotFound and Forbidden. Made redirection base on http status code. Add page for errors (instead console.log()).

**Review by serge30 — COMMENTED** [review]
@RazSKg Please resolve conflicts.

**Comment by RazSKg** [comment]
@serge30 ready for review

File: frontend/src/components/UserInfo/UserAvatarWithInfo.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        if (success === null){
+            return null;
+        }
+
+        if (needRedirection === true){
+            let redirection = redirectToSignUp(error);
+            return redirection;
+        }
+        else{
+            if(success ===true){
```
Comment (serge30) [inline frontend/src/components/UserInfo/UserAvatarWithInfo.js]: But why? What's wrong with `if(success) {...`?

File: frontend/src/components/UserMenu.jsx [hunk 15]
```diff
@@ -65,7 +65,8 @@ class UserMenu extends React.Component {
   render() {
 
     const { auth, token, role } = this.props.ctx;
-    const isOwner = role === "Owner";
+    console.log(role)
```
Comment (serge30) [inline frontend/src/components/UserMenu.jsx]: No debug loging in code please.

File: easyrest/scripts/json_helpers.py [hunk 16]
```diff
@@ -1,13 +1,30 @@
 """Script contains helper functions to work with json such as:
     wrap(): wraper to make response acording to style
+    date_time_normalize(): function, which normalize date_time object for proper json serialization
 """
 
+import datetime
+
+
+def date_time_normalize(date_time_object):
```
Comment (serge30) [inline easyrest/scripts/json_helpers.py]: I do not think that `date_time_normalize` is still used.

File: easyrest/models/__init__.py [hunk 17]
```diff
@@ -17,6 +17,7 @@
 from .tag import Tag
 from .tag_association import TagAssociation
 from .user import User
+
```
Comment (serge30) [inline easyrest/models/__init__.py]: What is the purpose of this change?

**Review by serge30 — APPROVED** [review]
Please make a fixes according to comments and merge.

File: easyrest/models/__init__.py (hunk 17 above)
Comment (RazSKg) [reply easyrest/models/__init__.py]: It`s about space? If yes - I accidentally added this.
